test/e2e/instance.go

```diff
@@ -0,0 +1,145 @@
+package e2e
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	mmv1beta "github.com/mattermost/mattermost-operator/apis/mattermost/v1beta1"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	mattermostInstanceCreationTimeout = 15 * time.Second
+	mattermostInstanceWaitTimeout     = 3 * time.Minute
+	mattermostInstanceGetTimeout      = 60 * time.Second
+	mattermostInstanceUpdateTimeout   = 60 * time.Second
+)
+
+// mattermostInstance defines a mattermost instance that is created in the test k8s cluster in order to ease the creation,
+// destruction and testing of the defined spec.
+// TODO: Ensure creation and deletion only happen once using sync.Once ? Not sure this is necessary since this is a testing
+// tool.
+type mattermostInstance struct {
+	t              *testing.T
+	mattermostSpec *mmv1beta.Mattermost
+	timeoutCreate  time.Duration
+	timeoutWait    time.Duration
+	timeoutGet     time.Duration
+	timeoutUpdate  time.Duration
+	k8sClient      client.Client
+	namespaceName  types.NamespacedName
+	created        bool
+}
+
+// Namespace returns the NamespacedName in order to retrieve the object easily
+func (m *mattermostInstance) Namespace() types.NamespacedName {
+	return m.namespaceName
+}
+
+// Create creates the instance within the cluster
+func (m *mattermostInstance) Create() {
+	m.namespaceName = types.NamespacedName{
+		Namespace: m.mattermostSpec.Namespace,
+		Name:      m.mattermostSpec.Name,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), m.timeoutCreate)
+	defer cancel()
+
+	err := m.k8sClient.Create(ctx, m.mattermostSpec)
+	require.NoError(m.t, err)
+
+	m.created = true
+}
+
+// CreateAndWait Creates the instance within the cluster and waits until is a stable instance (failing if not)
+func (m *mattermostInstance) CreateAndWait() {
+	m.Create()
+	m.Wait()
+}
+
+// Wait waits for the mattermost instance to be stable
+func (m *mattermostInstance) Wait() {
+	err := WaitForMattermostStable(m.t, m.k8sClient, m.Namespace(), m.timeoutWait)
+	require.NoError(m.t, err, "Timed out waiting for a stable mattermost instance")
+}
+
+// Destroy destroys the created instance
+func (m *mattermostInstance) Destroy() {
+	if !m.created {
+		return
+	}
+
+	err := m.k8sClient.Delete(context.Background(), m.mattermostSpec)
+	require.NoError(m.t, err)
```

Comment:
I think we can just not fail if error type is `NotFound` and remove the `created` variable altogether.

Reply:
Totally agree. Didn't like the `created` approach much since this is a developer testing tool, but I wanted to assure that we don't just fail to destroy an instance that wasn't created in the first place (though that "shouldn't happen"). Your approach is way better!